lib/collection/src/operations/types.rs

```diff
@@ -560,6 +574,69 @@ pub struct RecommendGroupsRequest {
     pub group_request: BaseGroupRequest,
 }
 
+/// Use context and a target to find the most similar points, constrained by the context.
+///
+/// When using only the context, a special search is performed where pairs of points are
+/// used to generate a loss that guides the search towards the zone where most positive
+/// examples overlap. This means that the score minimizes the scenario of finding a point
+/// closer to a negative than to a positive part of a pair.
+/// Since the score of a context relates to loss, the maximum score a point can get is 0.0,
+/// and it becomes normal that many points can have 0.0 as score.
+///
+/// Using only a target is equivalent to regular search, so the score is the distance to the target.
+///
+/// When using both context and target, the score behaves a little different: The
+/// integer part of the score represents the "rank" with respect to the context, while the
+/// decimal part of the score relates to the distance to the target.
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Validate, Clone)]
+pub struct DiscoverRequest {
+    /// Look for vectors closest to this
+    pub target: Option<RecommendExample>,
+
+    /// Pairs of (positive, negative) examples to provide context to the search
+    pub context_pairs: Option<Vec<(RecommendExample, RecommendExample)>>,
```

Comment:
Shall we drop pairs that have the exact same A/B vector?

We can probably do that in the search implementation with `pairs.filter(A != B)`.

We can also add validation here to prevent it, but I guess that might break the discovery API unexpectedly for some users.

Reply:
Good question, in the case of Discovery search, rank would be 0 if distance is the same against both pairs, so it should not have any adverse effect. For Context search, we add a very small margin so it may behave a little weird.

I think it is a good idea to cases in preprocessing 👍